TLS.md

```diff
@@ -0,0 +1,194 @@
+# How to Set Up TLS for NVMe-TCP
+
+Enabling TLS for the NVMe-TCP feature requires a few configuration steps for both the kernel and userland.
+
+## Kernel Configuration
+
+To support TCP authentication and TLS encryption, enable the following kernel options:
+
+- For DHCHAP authentication:
+  `CONFIG_NVME_HOST_AUTH`
+
+- For TLS transport encryption:
+  `CONFIG_NVME_TCP_TLS`
+
+## Setting Up `tlshd`
+
+For TLS protocol support, which handles authentication and encryption, the kernel handles data encryption only, so userland support is required for the TLS handshake. The `tlshd` daemon implements the handshake process.
+
+### Requirements
+
+Ensure `tlshd` includes the commit `311d9438b984` ("tlshd: always link .nvme default keyring into the session") - likely in `ktls-utils` version 0.12. Alternatively, you can set the keyring manually in `/etc/tlshd.conf`:
+
+```ini
+[authenticate]
+keyrings = .nvme
+```
+
+No additional configuration is necessary for `tlshd`; simply start it as a daemon:
+
+```bash
+systemctl enable --now tlshd
+```
+
+## Loading Keys on Boot or Module Load
+
+The NVMe subsystem loading keys from the kernel keystore, which means these keys must be available in the keystore before establishing a connection.
+
+### Creating a New Key
+
+```bash
+nvme gen-tls-key \
+  --hostnqn nqn.2014-08.org.nvmexpress:uuid:befdec4c-2234-11b2-a85c-ca77c773af36 \
+  --subsysnqn nqn.io-1 --hmac 1 --identity 1 --insert --keyfile /etc/nvme/tls-keys
+```
+
+This command creates a new host key, inserts it into the kernel keyring, and appends the derived TLS PSK to the keyfile (`/etc/nvme/tls-keys`).
+
+### Inserting an Existing Key
+
+```bash
+nvme check-tls-key \
+  --hostnqn nqn.2014-08.org.nvmexpress:uuid:befdec4c-2234-11b2-a85c-ca77c773af36 \
+  --subsysnqn nqn.io-1 --identity 1 \
+  --keydata NVMeTLSkey-1:01:AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACtVQoZ: \
+  --insert --keyfile /etc/nvme/tls-keys
+```
+
+This command inserts the configured key (`--keydata`) into the kernel keyring and appends the derived TLS PSK to the keyfile.
+
+### Loading Keys on Boot or Module Load
+
+The kernel keyring does not persist keys, so userland must import keys into the keyring upon each boot or module load (for NVMe-TCP). To load keys:
+
+```bash
+nvme tls --import /etc/nvme/tls-keys
```

Comment:
This will hang waiting for stdin input. This should probably be replaced with ```nvme tls --import --keyfile=/etc/nvme/tls-keys``` or ```nvme tls < /etc/nvme/tls-keys``` here.

Reply:
I should have copied these commands verbatim from my test script...